File: client/packagist.go
package client

import (
	"encoding/json"
	"io/ioutil"
	"log"
	"net/http"
	"strings"
	"time"
)

const packagesUrl = "https://packagist.org/"

type PackageList struct {
	PackageNames []string `json:"packageNames"`
}

type PackageDetail struct {
	Package struct {
		Name        string    `json:"name"`
		Description string    `json:"description"`
		Time        time.Time `json:"time"`
		Maintainers []struct {
			Name      string `json:"name"`
			AvatarURL string `json:"avatar_url"`
		} `json:"maintainers"`
		Versions struct {
			DevMaster struct {
				Name              string        `json:"name"`
				Description       string        `json:"description"`
				Keywords          []string      `json:"keywords"`
				Homepage          string        `json:"homepage"`
				Version           string        `json:"version"`
				VersionNormalized string        `json:"version_normalized"`
				License           []string      `json:"license"`
				Authors           []interface{} `json:"authors"`
				Source            struct {
					Type      string `json:"type"`
					URL       string `json:"url"`
					Reference string `json:"reference"`
				} `json:"source"`
				Dist struct {
					Type      string `json:"type"`
					URL       string `json:"url"`
					Reference string `json:"reference"`
					Shasum    string `json:"shasum"`
				} `json:"dist"`
				Type    string `json:"type"`
				Support struct {
					Source string `json:"source"`
					Issues string `json:"issues"`
				} `json:"support"`
				Time  time.Time `json:"time"`
				Extra struct {
					InstallerName       string `json:"installer-name"`
					ShopwarePluginClass string `json:"shopware-plugin-class"`
				} `json:"extra"`
				DefaultBranch bool `json:"default-branch"`
				Require       struct {
					Php                string `json:"php"`
					ComposerInstallers string `json:"composer/installers"`
				} `json:"require"`
			} `json:"dev-master"`
		} `json:"versions"`
		Type             string           `json:"type"`
		Repository       string           `json:"repository"`
		GithubStars      int              `json:"github_stars"`
		GithubWatchers   int              `json:"github_watchers"`
		GithubForks      int              `json:"github_forks"`
		GithubOpenIssues int              `json:"github_open_issues"`
		Language         string           `json:"language"`
		Dependents       int              `json:"dependents"`
		Suggesters       int              `json:"suggesters"`
		Downloads        PackageDownloads `json:"downloads"`
		Favers           int              `json:"favers"`
	} `json:"package"`
}

type PackageDownloads struct {
	Total   int `json:"total"`
	Monthly int `json:"monthly"`
	Daily   int `json:"daily"`
}

type PackageStatistics struct {
	Github struct {
		Stars    int
		Watchers int
		Forks    int
	}
	Downloads PackageDownloads
}

func request(url string) []byte {
	resp, err := http.Get(url)

	if err != nil {
		log.Fatal(err)
	}

	defer resp.Body.Close()

	body, err := ioutil.ReadAll(resp.Body)

	if err != nil {
		log.Fatal(err)
	}

	return body
}

func getPluginNameFromPackage(detail PackageDetail) string {
	if len(detail.Package.Versions.DevMaster.Extra.ShopwarePluginClass) > 0 {
		split := strings.Split(detail.Package.Versions.DevMaster.Extra.ShopwarePluginClass, "\\")
		return split[len(split)-1]
	}

	return detail.Package.Versions.DevMaster.Extra.InstallerName
}

func GetPackageStatistics() map[string]*PackageStatistics {
	var Packages = make(map[string]*PackageStatistics)

	var packageList PackageList
	err := json.Unmarshal(request(packagesUrl+"packages/list.json?vendor=frosh"), &packageList)
	if err != nil {
		log.Println(err)
	}

	for _, name := range packageList.PackageNames {
		var pkg PackageDetail
		err := json.Unmarshal(request(packagesUrl+"packages/"+name+".json"), &pkg)
		if err != nil {
			log.Println(err)
		}

		pluginName := getPluginNameFromPackage(pkg)

		if len(pluginName) == 0 {
			continue
		}

		Packages[pluginName] = &PackageStatistics{
			Downloads: pkg.Package.Downloads,
			Github: struct {
				Stars    int
				Watchers int
				Forks    int
			}{Stars: pkg.Package.GithubStars, Watchers: pkg.Package.GithubWatchers, Forks: pkg.Package.GithubForks},
		}
	}

	return Packages
}

File: src/services/packagist.ts
import { PackageStatistics } from '../types';

const PACKAGIST_URL = 'https://packagist.org/';

interface PackageList {
  packageNames: string[];
}

interface PackageDetail {
  package: {
    name: string;
    description: string;
    time: string;
    maintainers: Array<{
      name: string;
      avatar_url: string;
    }>;
    versions: {
      'dev-master': {
        name: string;
        description: string;
        keywords: string[];
        homepage: string;
        version: string;
        version_normalized: string;
        license: string[];
        authors: any[];
        source: {
          type: string;
          url: string;
          reference: string;
        };
        dist: {
          type: string;
          url: string;
          reference: string;
          shasum: string;
        };
        type: string;
        support: {
          source: string;
          issues: string;
        };
        time: string;
        extra: {
          'installer-name': string;
          'shopware-plugin-class': string;
        };
        'default-branch': boolean;
        require: {
          php: string;
          'composer/installers': string;
        };
      };
    };
    type: string;
    repository: string;
    github_stars: number;
    github_watchers: number;
    github_forks: number;
    github_open_issues: number;
    language: string;
    dependents: number;
    suggesters: number;
    downloads: {
      total: number;
      monthly: number;
      daily: number;
    };
    favers: number;
  };
}

export class PackagistService {
  private async request(url: string): Promise<any> {
    try {
      const response = await fetch(url);
      if (!response.ok) {
        throw new Error(`HTTP error! Status: ${response.status}`);
      }
      return await response.json();
    } catch (error) {
      console.error(`Error fetching ${url}:`, error);
      throw error;
    }
  }

  private getPluginNameFromPackage(detail: PackageDetail): string {
    const devMaster = detail.package.versions['dev-master'];
    if (devMaster?.extra?.['shopware-plugin-class']?.length > 0) {
      const split = devMaster.extra['shopware-plugin-class'].split('\\');
      return split[split.length - 1];
    }
    return devMaster?.extra?.['installer-name'] || '';
  }

  async getPackageStatistics(): Promise<Record<string, PackageStatistics>> {
    const packages: Record<string, PackageStatistics> = {};

    try {
      // Get list of packages
      const packageList: PackageList = await this.request(`${PACKAGIST_URL}packages/list.json?vendor=frosh`);
      
      // Process each package
      for (const name of packageList.packageNames) {
        try {
          const packageDetail: PackageDetail = await this.request(`${PACKAGIST_URL}packages/${name}.json`);
          const pluginName = this.getPluginNameFromPackage(packageDetail);
          
          if (pluginName.length === 0) {
            continue;
          }
          
          packages[pluginName] = {
            downloads: packageDetail.package.downloads,
            github: {
              stars: packageDetail.package.github_stars,
              watchers: packageDetail.package.github_watchers,
              forks: packageDetail.package.github_forks
            }
          };
        } catch (error) {
          console.error(`Error processing package ${name}:`, error);
        }
      }
      
      return packages;
    } catch (error) {
      console.error('Error fetching package statistics:', error);
      return {};
    }
  }
} 